goaci.go

```diff
@@ -47,189 +27,398 @@ func (v *StringVector) Set(str string) error {
+	exec := []string{filepath.Join("/", binary)}
+	exec = append(exec, opts.exec...)
+
+	return &types.App{
+		Exec:  exec,
+		User:  "0",
+		Group: "0",
 	}
-	fn := fi[0].Name()
-	debug("found binary: ", fn)
+}
 
-	// Set up rootfs for ACI layout
-	rfs := filepath.Join(acidir, "rootfs")
-	err = os.MkdirAll(rfs, 0755)
+func getVCSLabel(pathsNames *pathsAndNames) (*types.Label, error) {
+	name, value, err := GetVCSInfo(pathsNames.projectPath)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get VCS info: %v", err)
+	}
+	acname, err := types.NewACName(name)
 	if err != nil {
-		die(err.Error())
+		return nil, fmt.Errorf("Invalid VCS label: %v", err)
 	}
+	return &types.Label{
+		Name:  *acname,
+		Value: value,
+	}, nil
+}
 
-	// Move the binary into the rootfs
-	ep := filepath.Join(rfs, fn)
-	err = os.Rename(filepath.Join(gobin, fn), ep)
+func prepareManifest(opts *options, pathsNames *pathsAndNames, binary string) (*schema.ImageManifest, error) {
+	name, err := types.NewACName(pathsNames.imageACName)
+	// TODO(jonboulle): could this ever actually happen?
+	if err != nil {
+		return nil, err
+	}
+
+	app := getApp(opts, binary)
+
+	vcsLabel, err := getVCSLabel(pathsNames)
 	if err != nil {
-		die(err.Error())
+		return nil, err
+	}
+	labels := types.Labels{
+		*vcsLabel,
+	}
+
+	manifest := schema.BlankImageManifest()
+	manifest.Name = *name
+	manifest.App = app
+	manifest.Labels = labels
+	return manifest, nil
+}
+
+func copyAssets(opts *options, pathsNames *pathsAndNames) error {
+	paths := map[string]string{
+		"<PROJPATH>": pathsNames.projectPath,
+		"<GOPATH>":   pathsNames.goPath,
+	}
+	if err := PrepareAssets(opts.assets, pathsNames.rootFSPath, paths); err != nil {
+		return err
 	}
-	debug("moved binary to:", ep)
+	return nil
+}
+
+func moveBinaryToRootFS(pathsNames *pathsAndNames, binary string) error {
+	// Move the binary into the rootfs
+	ep := filepath.Join(pathsNames.rootFSPath, binary)
+	if err := os.Rename(filepath.Join(pathsNames.goBinPath, binary), ep); err != nil {
+		return err
+	}
+	Debug("moved binary to: ", ep)
+	return nil
+}
 
-	exec := []string{filepath.Join("/", fn)}
-	exec = append(exec, execOpts...)
-	// Build the ACI
-	im := schema.ImageManifest{
-		ACKind:    types.ACKind("ImageManifest"),
-		ACVersion: schema.AppContainerVersion,
-		Name:      *name,
-		App: &types.App{
-			Exec:  exec,
-			User:  "0",
-			Group: "0",
-		},
+func writeACI(pathsNames *pathsAndNames, manifest *schema.ImageManifest) error {
+	mode := os.O_CREATE | os.O_WRONLY | os.O_TRUNC
+	of, err := os.OpenFile(pathsNames.imageFileName, mode, 0644)
+	if err != nil {
+		return fmt.Errorf("Error opening output file: %v", err)
 	}
-	debug(im)
+	defer of.Close()
 
 	gw := gzip.NewWriter(of)
+	defer gw.Close()
+
 	tr := tar.NewWriter(gw)
+	defer tr.Close()
+
+	iw := aci.NewImageWriter(*manifest, tr)
+	if err := filepath.Walk(pathsNames.aciDirPath, aci.BuildWalker(pathsNames.aciDirPath, iw)); err != nil {
```

Comment:
[aci.BuildWalker](https://github.com/appc/spec/blob/master/aci/build.go#L12) adds files in the tarball and preserves the uid/gid of the files. I am testing ``goaci`` as user alban and uid 1000 so the files copied by goaci in the temporary directory belong to 1000 and in files in the tar archive belong to 1000 (see [populateHeaderUnix](https://github.com/appc/spec/blob/master/pkg/tarheader/pop_posix.go#L32)).

This goes against https://github.com/appc/goaci/issues/10. For now, I think we should just mark all files as belonging to root in the tar file. We can revisit this when https://github.com/appc/spec/issues/231 gets fixed.

It might be possible to mark all files as belonging to root without patching [aci.BuildWalker](https://github.com/appc/spec/blob/master/aci/build.go#L12) by adding an intermediary function call between Walk and BuildWalker which will provide a custom [os.FileInfo](http://golang.org/pkg/os/#FileInfo).

In my opinion, this can be done later: in this pull request, we can just add a FIXME comment with a link to an issue about that.